src/Workspaces/Remote/Core/Serialization/ImmutableCollectionMessagePackResolver.cs

```diff
@@ -0,0 +1,394 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Reflection;
+using MessagePack;
+using MessagePack.Formatters;
+
+// TODO: Copied from https://github.com/neuecc/MessagePack-CSharp/blob/master/src/MessagePack.ImmutableCollection/Formatters.cs.
+// Replace with an implementation shipping with VS:
+// https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1198374
+// https://github.com/neuecc/MessagePack-CSharp/issues/606
+//
+// The following code includes fix for: https://github.com/neuecc/MessagePack-CSharp/issues/1033. Make sure the fix is included in the replacement.
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    internal sealed class ImmutableCollectionMessagePackResolver : IFormatterResolver
```

Comment:
what's an IFormatterResolver?

Reply:
MessagePack's interface for authoring custom serializers.